source/2_platform/rex_directx/src/directx_renderer.cpp

```diff
@@ -73,76 +238,1816 @@ namespace rex
+        backend::release_resource(ls.vertex_shader); // vertex shader is referenced in the shader program
+        backend::release_resource(ls.pixel_shader);  // pixel shader is referenced in the shader program
+
+        return true;
+      }
+
+      //-------------------------------------------------------------------------
+      bool compile_shader(const commands::CompileShaderCommandDesc& cs, const ResourceSlot& resourceSlot)
+      {
+        s32 compile_flags = 0;
+#if defined(REX_DEBUG)
+        compile_flags = D3DCOMPILE_DEBUG | D3DCOMPILE_SKIP_OPTIMIZATION;
+#endif
+
+        HRESULT hr = S_OK;
+
+        wrl::com_ptr<ID3DBlob> byte_code = nullptr;
+        wrl::com_ptr<ID3DBlob> errors    = nullptr;
+
+        hr = D3DCompile2(cs.shader_code.data(), cs.shader_code.size(), cs.shader_name.c_str(), nullptr, D3D_COMPILE_STANDARD_FILE_INCLUDE, cs.shader_entry_point.data(), cs.shader_feature_target.data(), compile_flags, 0, 0, 0, 0, &byte_code, &errors);
+
+        if(errors != nullptr)
+        {
+          REX_ERROR(LogDirectX, "{}", (char*)errors->GetBufferPointer());
+          return false;
+        }
+
+        if(FAILED(hr))
+        {
+          REX_ERROR(LogDirectX, "Failed to compile shader");
+          return false;
+        }
+
+        switch(cs.shader_type)
+        {
+          case ShaderType::VERTEX: g_ctx.resource_pool.insert(resourceSlot, rsl::make_unique<VertexShaderResource>(byte_code)); break;
+          case ShaderType::PIXEL: g_ctx.resource_pool.insert(resourceSlot, rsl::make_unique<PixelShaderResource>(byte_code)); break;
+
+          default: REX_ERROR(LogDirectX, "Unsupported Shader Type was given"); return false;
+        }
+
+        return true;
+      }
+
+      //-------------------------------------------------------------------------
+      void update_commited_resource(const commands::UpdateCommitedResourceCommandDesc& updateConstantBuffer, const ResourceSlot& resourceSlot)
+      {
+        auto& cbr = get_resource_from_pool_as<CommitedBufferResource>(g_ctx.resource_pool, resourceSlot);
+        auto cs   = cbr.get();
+
+        memcpy(&cs->mapped_data[updateConstantBuffer.element_index * cs->element_data_byte_size], updateConstantBuffer.buffer_data.data(), updateConstantBuffer.buffer_data.size());
+      }
+
+      //-------------------------------------------------------------------------
+      void wait_for_active_frame()
+      {
+        // Cycle through the circular frame resource array.
+        g_ctx.frame_ctx.curr_frame_resource_index = (g_ctx.frame_ctx.curr_frame_resource_index + 1) % g_ctx.frame_ctx.frame_resources.size();
+        g_ctx.frame_ctx.curr_frame_resource       = &g_ctx.frame_ctx.frame_resources[g_ctx.frame_ctx.curr_frame_resource_index].slot;
+
+        auto& fr = get_resource_from_pool_as<FrameResource>(g_ctx.resource_pool, *g_ctx.frame_ctx.curr_frame_resource);
+        auto f   = fr.get();
+
+        if(f->fence != 0 && g_ctx.fence->GetCompletedValue() < f->fence)
```

Comment:
this code is the same as what you have in flush_command_queue
I suggest to move it to another func and call that func here and from flush_command_queue.
what you're doing is waiting for a fence

Reply:
It actually is not, this one is checking the fence of the active frame.
The one in flush command queue is the global fence.
This one also has additional checks.